### ISSUE TITLE
[GKE] Service of type LoadBalancer stays stuck if loadBalancerIP is static

### ISSUE THREAD

**Issue by riturajtiwari** [kubernetes/kubernetes#44521]
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):
No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):
"external ip not assigned on GKE " and variants
---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
BUG REPORT

**Kubernetes version** (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.0", GitCommit:"fff5156092b56e6bd60fff75aad4dc9de6b6ef37", GitTreeState:"clean", BuildDate:"2017-03-28T16:36:33Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.0", GitCommit:"fff5156092b56e6bd60fff75aad4dc9de6b6ef37", GitTreeState:"clean", BuildDate:"2017-03-28T16:24:30Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: Google Compute Platform

**What happened**:
Several of the details on this thread: [http://stackoverflow.com/questions/43416772/service-wont-come-up-when-using-static-ip-on-gke](http://stackoverflow.com/questions/43416772/service-wont-come-up-when-using-static-ip-on-gke).
TL;DR: Trying to create a service and expose it via static IP on GCP. If I assign a `Service` `LoadBalancer` with a `loadBalancerIP` that is the same as the static IP assigned on GCP, the service stays stuck in external IP allocation pending state. 

```
$ kubectl get service my-app-service
NAME                   CLUSTER-IP     EXTERNAL-IP   PORT(S)        AGE
my-app-service          10.x.x.x        <pending>     80:31432/TCP   12m
```

**What you expected to happen**:
Service will allocate itself the static IP I assigned

**How to reproduce it** (as minimally and precisely as possible):
- Allocate a static IP on GCP
- Use that static IP on a `Service` of type `LoadBalancer`

**Anything else we need to know**:
I think there is generally missing guidance on how to combine ingress with service, or maybe my perception is clouded by the errors I am seeing.

**Comment by cblecker** [comment]
This isn't a bug. If the static IP address is already in use (in this case, by the Ingress per your Stackoverflow link) or is the wrong type (LoadBalancer IPs in GCE need to be regional and in the same region as the cluster) then the LoadBalancer service will go into pending. Global IPs are used by Layer 7 Loadbalancer provisioned by the ingress.

/assign

**Comment by riturajtiwari** [comment]
Thanks @cblecker for the review. If I go from `LoadBalancer` to `NodePort` service, I still have a very long delay in connecting the `Ingress` to the `Service` - 10 minutes or so. 

Maybe I am not doing something right. Is there another recommended way of connecting `Ingress` to a `Service`?

**Comment by riturajtiwari** [comment]
Any update on this? This is clearly impaired functionality *somewhere*. I have done several experiments and consistently find that the `Ingress` takes several minutes to get connected to `Service`. What information can I give to take the next step here?

**Comment by cblecker** [comment]
Hi @riturajtiwari,
Sorry, I didn't notice the last message. I believe this provisioning time is related to the GCE Layer 7 Load Balancer, rather than Kubernetes itself.

cc: @kubernetes/goog-gke

**Comment by riturajtiwari** [comment]
Thanks @cblecker. One final question: Should I declare my service to be `LoadBalancer`, `NodePort` or whatever the default is? I am finding that if I use `NodePort`, things work after the several minute L7 delay, but if I don't use any type specification, things won't come up for over 30 minutes?

**Comment by cblecker** [comment]
In GKE:
- If you use a `LoadBalancer` type Service you provision a regional IP and Layer 4 load balancer that directs traffic to your application at the network layer.
- If you use an `Ingress` object, you provision the GCE Layer 7 HTTP Load Balancer. The should be paired with a `NodePort` service per the documentation here: https://cloud.google.com/container-engine/docs/tutorials/http-balancer#step_2_expose_nginx_as_a_service

**Comment by riturajtiwari** [comment]
Awesome! Thanks @cblecker !

**Comment by cblecker** [comment]
No problem!

/close

**Comment by riturajtiwari** [comment]
@cblecker The k8s web UI does not give any indication that the `Ingress` is not fully connected to the `Service`. Do you think this is a defect or a feature request? I am not sure how to best position this, hence asking. Thanks.

**Comment by ahmetb** [comment]
@riturajtiwari I don't think GCLB API provides any information on whether routing/forwarding rules are propagated to all its edge locations across the world. This probably makes Kubernetes not able to tell the readiness of the Ingress.

Keep in mind that GCLB is not regional and any configuration applied to it has to propagate all Google locations worldwide. More info at https://cloud.google.com/load-balancing/ That said, it'd be nice if there was a readiness indicator on the API, I agree.

**Comment by riturajtiwari** [comment]
@ahmetb Thanks for the comment. What is a good way to get Google's attention on this? I seem to be stuck in this ping pong where Google support essentially disowns anything k8s. They ask me to resolve with the k8s community.

**Comment by ahmetb** [comment]
@riturajtiwari I'm not sure what you mean. What I commented on is, Google Cloud Load Balancers simply may take a few minutes as the forwarding rules are propagated across the globe. This is not specific to Kubernetes. If you were to create a HTTP/s Load Balancer manually yourself, you would see similar delays.

**Comment by riturajtiwari** [comment]
Agreed. This is about getting one party - GCP or Kubernetes to own the user/developer experience and be responsible/accountable for it. Right now, I have a situation where I can create a `Service` and an `Ingress` and point the `Ingress` at the service. Now, for some indefinite period of time, the `Ingress` is not really connected because something somewhere is being provisioned and can take several minutes. As you will see in https://github.com/kubernetes/kubernetes/issues/45438, this does not complete for hours. Who should I turn to for visibility here? Is it a k8s enhancement request? GCP?

**Comment by cblecker** [comment]
@riturajtiwari As it's specific to the Google Cloud Load Balancer and the Google Container Engine implementation, you'll likely get more traction with GCP support as opposed to the open source Kubernetes project.

**Comment by nicksardo** [comment]
@riturajtiwari In #45438, you say that it never goes from 502 to 200.  Therefore, it's not fair to say it takes hours to provision.  It's more likely a configuration problem that is resulting in it never provisioning.  I'll continue discussion in that thread.
I've done a lot of testing on the ingress controller and have never seen it take longer than a few minutes, though I hear it can sometimes take up to ten minutes. 

We actually have Kubernetes e2e tests that spin up an ingress and wait a **maximum** of 15 minutes. If it doesn't provision within that time, the test fails. We're testing this quite frequently: https://k8s-testgrid.appspot.com/google-gke#gke-ingress
